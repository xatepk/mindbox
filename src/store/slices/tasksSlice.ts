import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { ITask } from '../../models/models';
import { v4 as uuidv4 } from 'uuid';

const initialState = [] as ITask[];

const todoSlice = createSlice({
  name: 'todos',
  initialState,
  reducers: {
    addTodo: {
      reducer: (state, action: PayloadAction<ITask>) => {
        state.push(action.payload);
      },
      prepare: (description: string) => ({
        payload: {
          id: uuidv4(),
          description,
          completed: false,
        } as ITask,
      }),
    },
    removeTodo(state, action: PayloadAction<string>) {
      const index = state.findIndex((todo) => todo.id === action.payload);
      state.splice(index, 1);
    },
    setTodoStatus(state, action: PayloadAction<{ completed: boolean; id: string }>) {
      const index = state.findIndex((todo) => todo.id === action.payload.id);
      state[index].completed = action.payload.completed;
    },
    clearCompleted(state, action: PayloadAction<boolean>) {
      const newToDos = state.filter((todo) => todo.completed === action.payload);
      state.splice(0, state.length, ...newToDos);
    },
  },
});

export const { addTodo, removeTodo, setTodoStatus, clearCompleted } = todoSlice.actions;
export default todoSlice.reducer;
