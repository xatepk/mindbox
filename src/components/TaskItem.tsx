import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTrashCan } from '@fortawesome/free-regular-svg-icons'
import { useDispatch } from 'react-redux';
import { removeTodo, setTodoStatus } from '../store/slices/tasksSlice';


interface TodoProps {
  id: string,
  description: string,
  completed: boolean
}

function TaskItem({ id, description, completed}: TodoProps) {
  const dispatch = useDispatch();

  return (
    <li className={`tasks__item ${completed ? 'tasks__item_active' : ''}`}>
      <input type="checkbox" id={`checkbox${id}`}
      className="tasks__item-input"
      onChange={() => dispatch(setTodoStatus({ completed: !completed, id: id }))}
      value={description}
      checked={completed} />
      <label htmlFor={`checkbox${id}`}></label>
      <span className={`tasks__item-desc ${completed ? 'tasks__item-desc-completed' : ''}`}>
        {description}
      </span>
      {!completed &&
      <span
      role="button"
      onClick={() => dispatch(removeTodo(id))}
      className='tasks__icon'>
        <FontAwesomeIcon icon={faTrashCan} />
      </span>}

    </li>
  );
}

export default TaskItem;
