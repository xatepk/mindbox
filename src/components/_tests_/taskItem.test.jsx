import { fireEvent, render, screen } from "@testing-library/react";
import * as reduxHooks from 'react-redux';
import { ITask } from "../../models/models";
import * as actions from '../../store/slices/tasksSlice';

import TaskItem from '../TaskItem';

jest.mock('react-redux');

const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');

describe('TaskItem', () => {
  it('should create Task', () => {
    mockedDispatch.mockResolvedValue(jest.fn());

    const view = render(
      <TaskItem id='123' description="Redux Toolkit" completed={false} />
    );

    expect(view).toMatchSnapshot();
  });

  it('should dispatch actions', () => {
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);

    const mockedToogleCompleted = jest.spyOn(actions, 'setTodoStatus');
    const mockedRemoveTask = jest.spyOn(actions, 'removeTodo');

    render(
      <TaskItem id='123' description="Redux Toolkit" completed={false} />
    );

    fireEvent.click(screen.getByRole('checkbox'));
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(mockedToogleCompleted).toHaveBeenCalledWith({"completed": true, "id": "123"});

    fireEvent.click(screen.getByRole("button"));
    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(mockedRemoveTask).toHaveBeenCalledWith('123');

  });

})
