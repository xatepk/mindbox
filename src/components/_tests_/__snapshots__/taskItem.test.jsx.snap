// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem should create Task 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="tasks__item "
      >
        <input
          class="tasks__item-input"
          id="checkbox123"
          type="checkbox"
          value="Redux Toolkit"
        />
        <label
          for="checkbox123"
        />
        <span
          class="tasks__item-desc "
        >
          Redux Toolkit
        </span>
        <span
          class="tasks__icon"
          role="button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-trash-can "
            data-icon="trash-can"
            data-prefix="far"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M160 400C160 408.8 152.8 416 144 416C135.2 416 128 408.8 128 400V192C128 183.2 135.2 176 144 176C152.8 176 160 183.2 160 192V400zM240 400C240 408.8 232.8 416 224 416C215.2 416 208 408.8 208 400V192C208 183.2 215.2 176 224 176C232.8 176 240 183.2 240 192V400zM320 400C320 408.8 312.8 416 304 416C295.2 416 288 408.8 288 400V192C288 183.2 295.2 176 304 176C312.8 176 320 183.2 320 192V400zM317.5 24.94L354.2 80H424C437.3 80 448 90.75 448 104C448 117.3 437.3 128 424 128H416V432C416 476.2 380.2 512 336 512H112C67.82 512 32 476.2 32 432V128H24C10.75 128 0 117.3 0 104C0 90.75 10.75 80 24 80H93.82L130.5 24.94C140.9 9.357 158.4 0 177.1 0H270.9C289.6 0 307.1 9.358 317.5 24.94H317.5zM151.5 80H296.5L277.5 51.56C276 49.34 273.5 48 270.9 48H177.1C174.5 48 171.1 49.34 170.5 51.56L151.5 80zM80 432C80 449.7 94.33 464 112 464H336C353.7 464 368 449.7 368 432V128H80V432z"
              fill="currentColor"
            />
          </svg>
        </span>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="tasks__item "
    >
      <input
        class="tasks__item-input"
        id="checkbox123"
        type="checkbox"
        value="Redux Toolkit"
      />
      <label
        for="checkbox123"
      />
      <span
        class="tasks__item-desc "
      >
        Redux Toolkit
      </span>
      <span
        class="tasks__icon"
        role="button"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash-can "
          data-icon="trash-can"
          data-prefix="far"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M160 400C160 408.8 152.8 416 144 416C135.2 416 128 408.8 128 400V192C128 183.2 135.2 176 144 176C152.8 176 160 183.2 160 192V400zM240 400C240 408.8 232.8 416 224 416C215.2 416 208 408.8 208 400V192C208 183.2 215.2 176 224 176C232.8 176 240 183.2 240 192V400zM320 400C320 408.8 312.8 416 304 416C295.2 416 288 408.8 288 400V192C288 183.2 295.2 176 304 176C312.8 176 320 183.2 320 192V400zM317.5 24.94L354.2 80H424C437.3 80 448 90.75 448 104C448 117.3 437.3 128 424 128H416V432C416 476.2 380.2 512 336 512H112C67.82 512 32 476.2 32 432V128H24C10.75 128 0 117.3 0 104C0 90.75 10.75 80 24 80H93.82L130.5 24.94C140.9 9.357 158.4 0 177.1 0H270.9C289.6 0 307.1 9.358 317.5 24.94H317.5zM151.5 80H296.5L277.5 51.56C276 49.34 273.5 48 270.9 48H177.1C174.5 48 171.1 49.34 170.5 51.56L151.5 80zM80 432C80 449.7 94.33 464 112 464H336C353.7 464 368 449.7 368 432V128H80V432z"
            fill="currentColor"
          />
        </svg>
      </span>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
